### PR TITLE
Improving Compile-Time Safety

### PR DESCRIPTION
**Pull request by mstrchrstphr**
# Improving Compile-Time Safety
## tl;dr

**This is a significant architectural change that aims to make the way we interact with entities, pools and systems safer.** This is a WIP and I've opened this PR for discussion.
### Strongly typed pools

``` c#
// Strongly typed pools are generated for you
CorePool pool = Pools.core;

// You can upcast to the new generic pool interface
IPool genericPool = Pools.core;
```
### Strongly typed pool entities

``` c#
// Pools return an object that conforms to an interface for that pool.
// This interface only contains methods relevant to the components of
// that pool. 
ICorePoolEntity entity = Pools.core.CreateEntity();

// This prevents you from adding/removing the wrong components.
entity.isMovable = true; // OK: MoveableComponent is part of this pool
entity.isRain = true;    // COMPILE ERROR: "isRain not found" RainCompoment is NOT part of this pool

// This also makes the autocompletion list much nicer
pool.CreateEntity()
  .AddPosition(0f, 0f)
  .AddHealth(100);

// You can upcast to the new generic entity interface
IEntity genericEntity = Pools.core.CreateEntity();
```
### Typed Reactive Systems

``` c#
// New typed reactive systems provide a greater level of safety
class TimeSystem : IReactiveSystem<ICorePoolEntity> {
  public void Execute(List<ICorePoolEntity> entities) {
  }
}
```
### New generated interfaces for each component

Credit: @mzaks for the suggestion in chat.

``` c#
// An interface is generated for each component.
// Below is the interface for PositionComponent.
public interface IPositionEntity : IEntity {
  PositionComponent position { get; }
  bool hasPosition { get; }
  IPositionEntity AddPosition(int newX, int newY);
  IPositionEntity ReplacePosition(int newX, int newY);
  IPositionEntity RemovePosition();
}

// Now we can use interface composition for typed reactive systems
interface IPhysicalEntity : IPositionEntity, IMassEntity, IVelocityEntity { }
class VelocitySystem : IReactiveSystem<IPhysicalEntity> {
  public void Execute(List<IPhysicalEntity> entities) {
    // Now we only have a subset of components to work with :)
  }
}
```
### Backwards compatible

These updates are 100% backwards compatible with the current version of Entitas. It will not affect existing code relying on Entitas and should have a minimal impact on ongoing development.

This work should not affect any existing discussions relating to optimization or architecture.
## Motivation

This work started from #136.
## Roadmap
- ~~Maintain backwards compatibility.~~
- Ensure single entities are handled correctly.
- Ensure flag entities are handled correctly.
- Investigate failing tests.
- Make sure line endings are consistent.

### CODE REVIEW

**Comment by mstrchrstphr** [comment]
My next step is trying to figure how to build + package Entitas for Unity. I've had trouble with this #138.

I'd also like to add tests for the new generated code. In general, testing code generators is usually very tedious and the tests are fragile. Do we have any procedures in place for updating test fixtures, etc?

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Windows
 Thumbs.db
+.vs
```
Comment (mstrchrstphr) [inline .gitignore]: Temp cache produced by VS 2015

File: Entitas/Entitas.csproj [hunk 2]
```diff
@@ -87,18 +91,5 @@
...
-    <Folder Include="Entitas\CodeGenerator\" />
-    <Folder Include="Entitas\CodeGenerator\Attributes\" />
-    <Folder Include="Entitas\CodeGenerator\Generators\" />
-    <Folder Include="Entitas\CodeGenerator\Interfaces\" />
-    <Folder Include="Entitas\CodeGenerator\Intermediate\" />
-    <Folder Include="Entitas\CodeGenerator\Providers\" />
-    <Folder Include="Entitas\Serialization\Blueprints\" />
-    <Folder Include="Entitas\Serialization\Configuration\" />
-  </ItemGroup>
+  <ItemGroup />
```
Comment (mstrchrstphr) [inline Entitas/Entitas.csproj]: I don't understand this diff. My VS project seems to compile fine. I'll investigate further.

File: Entitas/Entitas/CodeGenerator/Generators/ComponentExtensionsGenerator.cs [hunk 3]
```diff
@@ -135,7 +195,7 @@ public class ComponentExtensionsGenerator : IComponentCodeGenerator {
         public Entity Add$Name($typedArgs) {
             var component = CreateComponent<$Type>($Ids.$Name);
 $assign
-            return AddComponent($Ids.$Name, component);
+            return (Entity)AddComponent($Ids.$Name, component);
```
Comment (mstrchrstphr) [inline Entitas/Entitas/CodeGenerator/Generators/ComponentExtensionsGenerator.cs]: Because the `IEntity` interface declares `IEntity AddComponent(int index, IComponent component);` this needs to be casted. It will always succeed.

File: Entitas/Entitas/Entity.cs [hunk 4]
```diff
@@ -6,37 +6,18 @@ namespace Entitas {
...
         public bool isEnabled { get { return _isEnabled; } }
 
-        /// componentPools is set by the pool which created the entity and is used to reuse removed components.
-        /// Removed components will be pushed to the componentPool.
-        /// Use entity.CreateComponent(index, type) to get a new or reusable component from the componentPool.
-        /// Use entity.GetComponentPool(index) to get a componentPool for a specific component index.
         public Stack<IComponent>[] componentPools { get { return _componentPools; } }
 
-        /// The poolMetaData is set by the pool which created the entity and contains information about the pool.
-        /// It's used to provide better error messages.
```
Comment (mstrchrstphr) [inline Entitas/Entitas/Entity.cs]: I move all these comments to the interface so they can be maintained in just one location.

File: Entitas/Entitas/Entity.cs [hunk 5]
```diff
@@ -363,14 +342,14 @@ public partial class Entity {
         public int retainCount { get { return owners.Count; } }
 
         /// Returns all the objects that retain this entity.
-        public readonly HashSet<object> owners = new HashSet<object>();
-
+        public HashSet<object> owners { get { return _owners; } }
+        private readonly HashSet<object> _owners = new HashSet<object>();
```
Comment (mstrchrstphr) [inline Entitas/Entitas/Entity.cs]: Converted this to a property so it could be expressed in the interface.

File: Entitas/Entitas/Interfaces/IEntity.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Entitas {
+
+    public delegate void EntityChanged(Entity entity, int index, IComponent component);
+    public delegate void ComponentReplaced(Entity entity, int index, IComponent previousComponent, IComponent newComponent);
+    public delegate void EntityReleased(Entity entity);
```
Comment (mstrchrstphr) [inline Entitas/Entitas/Interfaces/IEntity.cs]: Before these types were declared inside the `Entity` class. Since interfaces can't declare types I moved them here. Should I put them in their own namespace or leave them in the global Entitas namespace?

File: Entitas/Entitas/Interfaces/IPool.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Entitas {
+
+    public delegate void PoolChanged(Pool pool, Entity entity);
+    public delegate void GroupChanged(Pool pool, Group group);
```
Comment (mstrchrstphr) [inline Entitas/Entitas/Interfaces/IPool.cs]: See note in `IEntity` about delegate types...

File: Entitas/Entitas/Pool.cs [hunk 8]
```diff
@@ -6,46 +6,19 @@ namespace Entitas {
...
-        /// It's used to provide better error messages.
         public PoolMetaData metaData { get { return _metaData; } }
-
-        /// Returns the number of entities in the pool.
         public int count { get { return _entities.Count; } }
-
-        /// Returns the number of entities in the internal ObjectPool for entities which can be reused.
         public int reusableEntitiesCount { get { return _reusableEntities.Count; } }
-
-        /// Returns the number of entities that are currently retained by other objects (e.g. Group, GroupObserver, ReactiveSystem).
```
Comment (mstrchrstphr) [inline Entitas/Entitas/Pool.cs]: Moved comments to interface.

File: Readme/Readme.csproj [hunk 9]
```diff
@@ -208,9 +222,5 @@
     </Compile>
   </ItemGroup>
   <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
-  <ItemGroup>
-    <Folder Include="Readme\" />
-    <Folder Include="Readme\Generated\" />
-    <Folder Include="Linked Libraries\" />
-  </ItemGroup>
+  <ItemGroup />
```
Comment (mstrchrstphr) [inline Readme/Readme.csproj]: Again, I'm very confused by this project file diff. What have I done? 😮

File: Entitas/Entitas/CodeGenerator/Generators/ComponentExtensionsGenerator.cs [hunk 10]
```diff
@@ -22,6 +22,10 @@ public class ComponentExtensionsGenerator : IComponentCodeGenerator {
 
         static string generateComponentExtension(ComponentInfo componentInfo) {
             var code = addNamespace();
+            if (componentInfo.isSingleEntity == false) {
+                code += addEntityInterface(componentInfo);
+            }
+
```
Comment (mstrchrstphr) [inline Entitas/Entitas/CodeGenerator/Generators/ComponentExtensionsGenerator.cs]: Atm, I'm not generating interfaces for single entities. I'll remove this constraint soon.

File: Entitas/Entitas/CodeGenerator/Generators/PoolEntityGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                            component.generateComponent,
+                            component.generateMethods,
+                            component.generateIndex
+                        );
+                    });
+                    acc.AddRange(components);
+                    return acc;
+                })
+                .GroupBy(component => component.pools[0])
+                .Where(group => group.Key.Length > 0)
```
Comment (mstrchrstphr) [inline Entitas/Entitas/CodeGenerator/Generators/PoolEntityGenerator.cs]: After the pool components have been separated into groups according to their pool name,

e.g.

```c#
{
    "Core" : {
        // core components...
    },
    "Meta" : {
        // meta components...
    },
    // etc.
}
```

I filter to make sure there are no unnamed pools. I believe it is possible to make an unnamed pool like so:

```c#
[Pool]
public class SomeComponent : IComponent { }
```
Am I correct to be filtering these out or should I call the pool something like "Default"? In that case, what is the difference between an unnamed pool and having no pool? Does it simplify things to only generate these interfaces for components that have explicit pools?

Comment (mstrchrstphr) [inline Entitas/Entitas/CodeGenerator/Generators/PoolEntityGenerator.cs]: Just saw this :) https://github.com/sschmid/Entitas-CSharp/blob/d7fd022597516df36194ffa02fc77c77f37d6f51/Entitas/Entitas/CodeGenerator/Attributes/PoolAttribute.cs#L9

**Comment by mstrchrstphr** [comment]
Hey @sschmid, @mzaks, I was planning on continuing work on this PR this week but before I do I wanted to get some feedback from you two: **Does this fit in the vision you have for Entitas? Are there plans that this interferes with?**

**Comment by sschmid** [comment]
Hi, having specific types to improve code completion will totally fit into the vision of Entitas. Once you scale your game you'll usually end up having multiple pools. This can lead to bugs because you are able to add components to entities from different pools etc. I'm currently integrating the new Pools class which is not static anymore and try to prepare a new version with features and bug fixes from all of you guys. After this I think it totally makes sense to keep on working on your suggestion.

**Comment by mstrchrstphr** [comment]
Good to hear. Thanks!

**Comment by sschmid** [comment]
Just a heads up: I really like your idea, great work! This topic just became my prio 1 task now. I will look more into this. I also added a new PoolsGenerator on develop for the new Pools instance class to make it easier to work with multiple pools. I also plan to replace pool.CreateSystems() with pools.CreateSystem() and move the responsibility to choose the correct pool into the reactive system itself. This all should fit nicely with all of your suggestions.

**Comment by sschmid** [comment]
I started to look into this now. I took a look at your changes and put them ontop of the latest develop.
I currently have problems getting it to work. Here's what I've done.

1. Added `IEntity`. Entity implements `IEntity`
2. Added `IPool<TEntity>`. Pool implements `IPool<TEntity>`
3. Updated Matcher, Group, GroupObserver, ReactiveSystem, TriggerOnEvent and more to be generic with TEntity
4. Added convenience default Matcher, Group, GroupObserver, ReactiveSystem, TriggerOnEvent with `<Entity>` to make VisualDebugging compile without changing too much.

I have questions:
Imagine TEntity is CoreEntity. What has to be generated and how should I use it?

Doing this
```csharp
public interface ICoreEntity : IEntity {
    PositionComponent position { get; }
    bool hasPosition { get; }
    Entity AddPosition(UnityEngine.Vector3 newPosition);
    Entity ReplacePosition(UnityEngine.Vector3 newPosition);
    Entity RemovePosition();
}
```

doesn't work with pool.CreateEntity() because new Entity cannot be casted to TEntity.
```csharp
public virtual TEntity CreateEntity() {
    var entity = _reusableEntities.Count > 0 ? _reusableEntities.Pop() : new Entity();
    ...
}
```

I can add constraints like this:
```csharp
public partial class Pool<TEntity> : IPool<TEntity> where TEntity : class, IEntity, new() { }
```

so I can do
```csharp
public virtual TEntity CreateEntity() {
    var entity = _reusableEntities.Count > 0 ? _reusableEntities.Pop() : new TEntity();
    ...
}
```

Then I have to create a pool like this
```csharp
var pool = new Pool<CoreEntity>(...);
```

Therefore I need a CoreEntity class
```csharp
public class CoreEntity : Entity { }
```

But by inheriting from Entity I basically get code completion for all pools.

So the solution would be
```csharp
public partial class CoreEntity : Entity { }
```

and the ComponentExtensionsGenerator should only add methods to CoreEntity and not Entity.
Is that correct? Did I miss sth?

**Comment by sschmid** [comment]
fyi, https://github.com/sschmid/Entitas-CSharp/commits/feature/type-safety

**Comment by sschmid** [comment]
What do you think of this idea:
- No interfaces for typed entities (e.g. CoreEntity)
- We keep `IEntity`, `IPool<TEntiy>`
- Entity is not partial anymore. Code Generator doesn't extend existing Entity class anymore
- Code Generator creates new partial class based on pool names, e.g. MetaEntity, CoreEntity
- Code Generator extends CoreEntity the same way it used to extend the normal Entity

```csharp
using Entitas;

public partial class CoreEntity : Entity {

    public PositionComponent position { get { return (PositionComponent)GetComponent(CoreComponentIds.Position); } }

    public bool hasPosition { get { return HasComponent(CoreComponentIds.Position); } }

    public CoreEntity AddPosition(UnityEngine.Vector3 newValue) {
        var component = CreateComponent<PositionComponent>(CoreComponentIds.Position);
        component.value = newValue;
        AddComponent(CoreComponentIds.Position, component);
        return this;
    }

    public CoreEntity ReplacePosition(UnityEngine.Vector3 newValue) {
        var component = CreateComponent<PositionComponent>(CoreComponentIds.Position);
        component.value = newValue;
        ReplaceComponent(CoreComponentIds.Position, component);
        return this;
    }

    public CoreEntity RemovePosition() {
        RemoveComponent(CoreComponentIds.Position);
        return this;
    }
}
```

- We used the new Pool `<TEntity>` as planned
```csharp
var corePool = new Pool<CoreEntity>(CoreComponentIds.TotalComponents);
corePool.CreateEntity()
        .AddPosition(new Vector3(1, 2, 3));
```

**Comment by sschmid** [comment]
Here's a working example
[Typed Entities.zip](https://github.com/sschmid/Entitas-CSharp/files/446745/Typed.Entities.zip)

**Comment by cloudjubei** [comment]
Looking at the proposed change (which I like and wanted to have typed pools for a very long time) there's a couple things I wanted to highlight that we wanted this to solve:
1) not having to write Add(Pools.core.CreateSystem<TypeOfSystem>()) or Add(Pools.core.CreateSystem(new TypeOfSystem())) in a Feature
 -> after typed pools writing Add(new TypeOfSystem()) and Entitas resolves the type based on trigger
2) if 1) is true then all Matchers need to be typed and I shouldn't have to specify anything other than the matcher (trigger) to communicate to the outside that a given ReactiveSystem works on a given Pool
3) Because we have typed Entities then only the components that belong to that given type are available on it from auto-complete -> otherwise compile-time error
4) as a bonus (dunno if you want to incorporate it into this feature) the static Pools class needs to go and either the ISetPools or some other idea needs to proposed so that ISystems will have access to the Pools without me having to write 3-5 lines everytime to get them (atm I solve by having an ASystem class that simply gets the non-static Pools in the constructor and stores it in a field)

So looking at Simon's latest commit 2) is not there and the question still remains whether 4) should be part of this already.

**Comment by sschmid** [comment]
1) my current plan is to have this API
```
pools.CreateSystem(new MySystem()));
```
Haven't check yet, but the idea is that you can infer the TEntity type of the system itself.

2) Yes, matchers should be typed as well
3) Definitely
4) The static Pools is already gone on develop. You can always write your own ISystem storing pools. I don't think this will be part of Entitas. You should use ISetPools instead of constructor arguments, otherwise you cannot easily reinitialize if you ever plan to do some kind of soft-reset

**Comment by cloudjubei** [comment]
1) this is an improvement on the current API but would be great if in a Feature I could cleanly write:
Add(new MySystem()); - plain and simple
4) ok cool - I guess that's good enough :)

**Comment by sschmid** [comment]
I'm currently playing around on this branch
https://github.com/sschmid/Entitas-CSharp/commits/feature/type-safety

**Comment by sschmid** [comment]
Here's an updated sample project
[Typed-Entities.zip](https://github.com/sschmid/Entitas-CSharp/files/450029/Typed-Entities.zip)

```csharp
Pools.sharedInstance.SetAllPools();

Pools.sharedInstance.meta.CreateEntity()
        .AddName("Max");

Pools.sharedInstance.core.CreateEntity()
        .AddPosition(new Vector3(1, 2, 3));
```
```csharp
[Core]
public class PositionComponent : IComponent {
    public Vector3 value;
}

[Meta]
public class NameComponent : IComponent {
    public string value;
}
```

You cannot add Position to a meta entity 👍

**Comment by sschmid** [comment]
The new Pools class let's you assign Pools as you like. It can store all specified pools. You can also set them to null now. It's planned to pass around the Pools instance with ISetPools instead of manually passing all required pool instances to an object.

**Comment by mstrchrstphr** [comment]
Whoa lots of new chatter 👍 

Thanks for moving this forward. Here are my thoughts on a few things you brought up:

>So the solution would ... the ComponentExtensionsGenerator should only add methods to [the Entity subclasses (which are partial)] and not Entity. Is that correct?

If that is the only thing we can do to make the type system happy then I'm good with that.

> No interfaces for typed entities (e.g. CoreEntity)

Even if the generated interfaces aren't used internally in Entitas I still think they are useful for doing things like composing interfaces for reactive systems:

```c#
// Now we can use interface composition for typed reactive systems
interface IPhysicalEntity : IPositionEntity, IMassEntity, IVelocityEntity { }
class VelocitySystem : IReactiveSystem<IPhysicalEntity> {
  public void Execute(List<IPhysicalEntity> entities) { }
}
```

**Comment by sschmid** [comment]
I feel like I am missing sth in your idea. I don't understand how it should work with the interfaces. Can you explain?
E.g. IPositionEntity, would you do sth like
```
var posEntity = (IPositionEntity)entity;
```
What type is entity? `Entity` or `CoreEntity`, how can you cast `Entity` to `IPositionEntity`?
Afaik, `Entity` would have to implement all those interfaces?

**Comment by sschmid** [comment]
It might work if the code generator makes the partial Entity class implement all interfaces, right?

**Comment by sschmid** [comment]
From my test, you cannot cast IEntity, ICoreEntity or anything else to IPhysicalEntity if you never declare interface ICoreEntity : IPhysicalEntity
Am I missing sth?

**Comment by sschmid** [comment]
My test setup:
```csharp

// given
public partial interface ICoreEntity : IEntity {}
public partial class CoreEntity : Entity, ICoreEntity {}

// having
[Pool("Core")]
public class NameComponent : IComponent {
    public string value;
}

// This will be generated
public interface ICoreNameRead {
    NameComponent name { get; }
    bool hasName { get; }
}

public interface ICoreNameWrite {
    void AddName(string value);
    void ReplaceName(string value);
    void RemoveName();
}

public partial class CoreEntity : ICoreNameRead, ICoreNameWrite {
    // placeholder for real index
    const int nameIndex = 0;

    public NameComponent name { get { return (NameComponent)GetComponent(nameIndex); } }

    public bool hasName { get { return HasComponent(nameIndex); } }

    public void AddName(string value) {
        var component = CreateComponent<NameComponent>(nameIndex);
        component.value = value;
        AddComponent(nameIndex, component);
    }

    public void ReplaceName(string value) {
        var component = CreateComponent<NameComponent>(nameIndex);
        component.value = value;
        AddComponent(nameIndex, component);
    }

    public void RemoveName() {
        RemoveComponent(nameIndex);
    }
}
```

```csharp
public interface Person : ICoreEntity, ICoreNameRead, ICoreNameWrite {}
```

```csharp
var pool = new Pool<CoreEntity>(...);
var e = (Person)pool.CreateEntity(); // Does not compile ;(
```

**Comment by sschmid** [comment]
Am I missing sth? It would be awesome if that would work

**Comment by mstrchrstphr** [comment]
Ok, so I don't think we'll be able to do that with C#. I was originally thinking we'd be able to do something in C# like this:

```swift
// Swift
protocol IMass { }
protocol IVelocity { }
protocol IPosition { }

typealias IPhysical = protocol<IMass, IVelocity, IPosition>
```

but that doesn't seem possible.

What if in order to get that behavior we did something like this:

```c#
// User-created interface
public interface IPerson : ICoreEntity, ICoreNameRead, ICoreNameWrite { }

// Make partial class conform to new interface.
// No need to implement anything since it already conforms.
public partial class CoreEntity : Person { } 

// Then...
var pool = new Pool<CoreEntity>(...);
var e = pool.CreateEntity() as IPerson;
```

**Comment by cloudjubei** [comment]
I personally think that the amount of code you as a dev have to write in your project (not in Entitas!) to get something like:
```
class VelocitySystem : IReactiveSystem<IPhysicalEntity> {
  public void Execute(List<IPhysicalEntity> entities) {...}
}
```
working, outweighs any usability this provides - as what it really only provides is code completion and ensuring that an entity has the given components, but this can be achieved already with the IEnsure/IExlude interfaces.

What I think you seem to want is actually a 'node' based approach (similar to Ash) where a system doesn't iterate over entities but over nodes which hold sets of components (for instance your Velocity, Mass, Position) and thus you can only interact with those in a given system. Simon and I have been talking about something like this for some time now and we definitely cannot have this real node based approach (where a node is a class container created every time and holding components) as it would be too slow, but we want to have the option of grouping components on a pool which would improve readability and performance, and hopefully at some point would lead us to a sort of interface like you have above :)

**Comment by mzaks** [comment]
This is what I did in the ECS-Lang code generator
I define the accessors as interfaces inside of the component, as I want to have an accessor based lens
```
public class JumpInTimeComponent : IComponent {
	public long value;
	public interface Get {JumpInTimeComponent jumpInTime{get;}}
	public interface Has {bool hasJumpInTime{get;}}
	public interface Add {void AddJumpInTime_(long newValue);}
	public interface Replace {void ReplaceJumpInTime_(long newValue);}
	public interface Remove {void RemoveJumpInTime_();}
}
```
Now I can define a new interface which aggregates those interfaces:
```
public interface InputEntity : JumpInTimeComponent.Get, JumpInTimeComponent.Add{}
```

And than I have to redefine partial Entity class that it implements the interface:
```
namespace Entitas {
	using my.game;
	public partial class Entity : ReplayBaseSystem.InputEntity {}
}
```

This works.

**Comment by cloudjubei** [comment]
Of course this works - but it means that the partial class Entity will implement a bazillion interfaces once you start making your own interface aggregates - dunno if there's really a problem with that. But more importantly, as I said above, I don't see any use for it other than code-completion.

**Comment by mzaks** [comment]
Well in case of the DSL it is a bit more than just autocompletion. It is a contract of how which data this system has access for and can create. The beauty of ECS is that everything is open and I don't want to attack this. I want to takle the problem that it is sometimes hard to see what is going on. With a definition of system contract we expose the data access and mutation on a higher level, without need to understand how it is done we see that it will access following components, replace this component and destroy this type of entity. Again let's have a Replay system as an example:

```
sys Replay {
    input unique {
        trigger : enteredGroup(JumpInTime)
        ensure : JumpInTime
        api : JumpInTime(get)
    }
    unique : ConsumtionHistory(has get)
    unique : LogicSystems(get)
    unique : Tick(replace)
    unique : JumpInTime(get)
    create consumeElixir : ConsumeElixir
}
```

On the first glance we directly see that it does has dependency to a lot of unique components, most of the dependencies are how ever read. Only Tick will be replaced, and we see that we will create a new Entity wich will have `ConsumeElixir` component. I believe it adds a great amount of value for behaviour analysis.

The generated interfaces just make sure that we don't cheat in the implementation. We have to stick with the described contract. I also don't believe that the overhead of describing the contract first will great, when you stick to the single responsibility principal.

**Comment by mstrchrstphr** [comment]
> It is a contract of - which data this system has access for and can create... The generated interfaces just make sure that we don't cheat in the implementation. We have to stick with the described contract.

Yes 👍  I think that we *could* achieve this with the IEnsure/IExlude interfaces, etc but that doesn't prevent the system from breaking the contract and modifying entities in whatever way it chooses. 

Additionally, as a project grows in size anything I can do to help my code write itself is very desirable. If a system only exposes the correct subset of an entity's interface to me (e.g. `IPerson` in my example above) then I am prevented by the compiler from making a logical error which I think is pretty neat. For example, I don't want a MoveSystem modifying an object's mass even though mass, velocity and position are likely components in the same pool.

**Comment by sschmid** [comment]
Heads up,
there's a version on this branch for everyone who wants to try it already.
https://github.com/sschmid/Entitas-CSharp/tree/feature/type-safety

This feature is currently on hold in favor of research for a roslyn based code generator.
But the type-safe feature will definitely be a thing in future versions of Entitas!

**Comment by sschmid** [comment]
Closed by #266
